File: node/subSets.js
const subSets = (arr) => {
  const results = []

  for (const item of arr) {
    const members = [...results]

    for (const s_item of members) {
      if (s_item !== null) results.push(item + s_item)
    }

    results.push(item)
  }

  return results
}

module.exports = { subSets }

describe("Set", () => {
  it("adds subsets", () => {
    expect(subSets([1])).toEqual([1])
    expect(subSets([1, 2]).sort()).toEqual([1, 2, 3].sort())
    expect(subSets([1, 2, 3]).sort()).toEqual([1, 2, 3, 3, 4, 5, 6].sort())
  })
})

File: ruby/sub_sets.rb
def sums(arr)
  results = []

  arr.each do |item|
    # can't add to the same members while iterating it, so we dup it
    members = results.dup
    members.each { |s_item| results.push(item + s_item) if s_item }

    results.push(item)
  end

 results
end

describe "Subsets" do
  it "returns all possible sum combinations given array of integers" do
    expect(sums([1])).to eq([1])
    expect(sums([1,2]).sort).to eq([1,2,3])
    expect(sums([1,2,3]).sort).to eq([1,2,3,3,4,5,6])
  end
end